From 9dff43ed2af6340cc9128440f7c95bc652896708 Mon Sep 17 00:00:00 2001
From: Robert Dabrowski <robertx.dabrowski@intel.com>
Date: Tue, 5 May 2020 12:37:15 +0200
Subject: [PATCH] PONRTSYS-6243: LAN TX drop counters not counting QoS drops -
 calculate TX drop counter subtracting CTP Bypass TX good packets from   CTP
 TX good packets

---
 .../net/datapath/mib_counters/dp_mib_counters.c    | 28 ++++++++++++++++++----
 1 file changed, 23 insertions(+), 5 deletions(-)

diff --git a/drivers/net/datapath/mib_counters/dp_mib_counters.c b/drivers/net/datapath/mib_counters/dp_mib_counters.c
index 574b3f2cad0c..d3c5307d6802 100644
--- a/drivers/net/datapath/mib_counters/dp_mib_counters.c
+++ b/drivers/net/datapath/mib_counters/dp_mib_counters.c
@@ -117,17 +117,35 @@ static void update_dev_counters(struct mib_counters *cnt, u8 initial_update)
 			return;
 		}
 	}
-
-	/* Overwrite CTP Tx counters by CTP Tx Pce-Bypass counters.
+	/* After this loop tmp_counters[0] contains all CTP counters, RX and TX
+	 * where RX counters are valid but TX can be not valid. 
+	 * In tmp_counters[1] there are CTP Bypass TX counters, all RX counters
+	 * here are invalid.
 	 * CTP Tx counters are placed before QoS modules, and they
 	 * record packets dropped by QoS modules too.
 	 * CTP Tx PCE-Bypass counters are placed after QoS modules,
 	 * and they record the packets actually transmitted.
-	 *
-	 * Do not overwrite CTP Tx counters for an UNI interface
-	 * in "shortcut" mode
+	 */
+
+	/* Overwrite CTP Tx counters by CTP Tx Pce-Bypass counters,
+	 * but do not overwrite CTP Tx counters for an UNI interface
+	 * in "shortcut" mode.
 	 */
 	if (!(cnt->dev_is_uni && shortcut_mode)) {
+	/* This check excludes overwriting CTP Counters by CTP Bypass Counters
+	 * for LAN ports if Shortcut mode is enabled.
+	 * In the same situation we do not expect any TX drop because
+	 * QOS doesn't filter the packets in Shortcut mode.
+	 *
+	 * For GEMs dev_is_uni is false so the drop counter will be calculated
+	 * for all gems. For LAN the code checks if there is Shortcut mode
+	 * enabled - if not, drop counter will be calculated.
+	 */
+
+		/* calculate QOS drop counter */
+		tmp_counters[1].nTxDroppedPkts = tmp_counters[0].nTxGoodPkts;
+		tmp_counters[1].nTxDroppedPkts -= tmp_counters[1].nTxGoodPkts;
+
 #define _DP_ETHTOOL_TX(r)	tmp_counters[0].r = tmp_counters[1].r;
 		DP_ETHTOOL_STAT_REG_TX_DECLARE
 #undef _DP_ETHTOOL_TX
